File: code/C#/Typeclasses/Concept/Ord.cs
namespace Typeclasses.Concept
{
    public interface Ord<in T>
    {
        bool gt(T @this, T that);
        bool eq(T @this, T that);
    }

    public class IntOrd : Ord<int>
    {
        public bool gt(int @this, int that) => @this > that;
        public bool eq(int @this, int that) => @this == that;
    }
    
    public class StringOrd : Ord<string>
    {
        public bool gt(string @this, string that) => string.CompareOrdinal(@this, that) > 0;
        public bool eq(string @this, string that) => string.CompareOrdinal(@this, that) == 0;
    }
}

File: code/Scala/src/main/scala/net/sigusr/concept/Ord.scala
package net.sigusr.concept

sealed trait Ord[-T] {
  def gt(l: T, r: T): Boolean
  def eq(l: T, r: T): Boolean
}

object IntOrd extends Ord[Int] {
  override def gt(l: Int, r: Int): Boolean = l > r
  override def eq(l: Int, r: Int): Boolean = l == r
}

object StringOrd extends Ord[String] {
  override def gt(l: String, r: String): Boolean = l > r
  override def eq(l: String, r: String): Boolean = l == r
}

